src/gui/fspathedit.cpp

```diff
@@ -0,0 +1,361 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "fspathedit.h"
+
+#include <QAction>
+#include <QApplication>
+#include <QFileDialog>
+#include <QHBoxLayout>
+#include <QStyle>
+#include <QToolButton>
+
+#include "base/utils/fs.h"
+#include "fspathedit_p.h"
+
+namespace
+{
+    struct TrStringWithComment
+    {
+        const char *source;
+        const char *comment;
+
+        QString tr() const
+        {
+            return QObject::tr(source, comment);
+        }
+    };
+
+    constexpr TrStringWithComment browseButtonBriefText =
+        QT_TRANSLATE_NOOP3("FileSystemPathEdit", "...", "Launch file dialog button text (brief)");
+    constexpr TrStringWithComment browseButtonFullText =
+        QT_TRANSLATE_NOOP3("FileSystemPathEdit", "&Browse...", "Launch file dialog button text (full)");
+    constexpr TrStringWithComment defaultDialogCaptionForFile =
+        QT_TRANSLATE_NOOP3("FileSystemPathEdit", "Choose a file", "Caption for file open/save dialog");
+    constexpr TrStringWithComment defaultDialogCaptionForDirectory =
+        QT_TRANSLATE_NOOP3("FileSystemPathEdit", "Choose a folder", "Caption for directory open dialog");
+}
+
+class FileSystemPathEdit::FileSystemPathEditPrivate
+{
+    Q_DECLARE_PUBLIC(FileSystemPathEdit)
+    Q_DISABLE_COPY(FileSystemPathEditPrivate)
+
+    FileSystemPathEditPrivate(FileSystemPathEdit *q, Private::FileEditorWithCompletion *editor);
+
+    void modeChanged();
+    void browseActionTriggered();
+    QString dialogCaptionOrDefault() const;
+
+    FileSystemPathEdit *q_ptr;
+    QScopedPointer<Private::FileEditorWithCompletion> m_editor;
+    QAction *m_browseAction;
+    QToolButton *m_browseBtn;
+    QString m_fileNameFilter;
+    Mode m_mode;
+    QString m_lastSignaledPath;
+    QString m_dialogCaption;
+};
+
+FileSystemPathEdit::FileSystemPathEditPrivate::FileSystemPathEditPrivate(FileSystemPathEdit *q,
```

Comment:
IMO, it would be better if you insert line break before first param (instead of second one).